torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp

```diff
@@ -196,6 +196,40 @@ class ConstCheck : private OptOutConstDispatch {
 
 } // namespace
 
+bool Val::sameAs(const Statement* other) const {
+  if (this == other) {
+    return true;
+  }
+  if (auto other_val = dynamic_cast<const Val*>(other)) {
+    if (definition_ == nullptr || other_val->definition_ == nullptr) {
+      return false;
+    }
+    if (vtype_ != other_val->vtype_) {
+      return false;
+    }
+    if (dtype_ != other_val->dtype_) {
+      return false;
+    }
+    if (!definition_->sameAs(other_val->definition_)) {
```

Comment:
Tested with
```C++
TEST_F(NVFuserTest, TEST) {
  auto fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  auto u = IrBuilder::create<Val>(ValType::Scalar);
  IrBuilder::create<UnaryOp>(UnaryOpType::Set, u, u);
  auto v = IrBuilder::create<Val>(ValType::Scalar);
  IrBuilder::create<UnaryOp>(UnaryOpType::Set, v, v);
  v->sameAs(u);
}
```
on my system, I can go depth >70k before hitting a segfault. For our indexing math, the depth of the expressions should be the depth of our transformations (not the number of transformations!), so I think we are pretty safe here.

Reply:
Thank you, that's really good to know. I know XLA suffered with really shallow traversal depth at one point so I was always a bit worried we would somehow get into trouble.

Reply:
I look forward to seeing a kernel with 70k indexing expressions in one line 🤣

Reply:
Haha, it actually can be higher than 70k, maybe much higher. I run the test and use gdb to print the backtrace when hitting a segfault. I am currently at 74k, but my screen is still scrolling down. Don't know how long it will take to finish scrolling.

Reply:
Aha, it stopped at level `174604`